capsules/src/virtual_rng.rs

```diff
@@ -0,0 +1,130 @@
+// Virtualizer for the RNG
+use core::cell::Cell;
+use kernel::common::cells::OptionalCell;
+use kernel::common::{List, ListLink, ListNode};
+use kernel::hil::rng::{Client, Continue, Rng};
+use kernel::ReturnCode;
+
+#[derive(Copy, Clone, PartialEq)]
+enum Op {
+    Idle,
+    Get,
+    SetClient,
+}
+
+// Struct to manage multiple rng requests
+pub struct MuxRngMaster<'a, R: Rng<'a>> {
+    rng: &'a dyn Rng<'a>,
+    devices: List<'a, VirtualRngMasterDevice<'a, R>>,
+    inflight: OptionalCell<&'a VirtualRngMasterDevice<'a, R>>,
+}
+
+impl<'a, R: Rng<'a>> MuxRngMaster<'a, R> {
+    pub const fn new(rng: &'a dyn Rng<'a>) -> MuxRngMaster<'a, R> {
+        MuxRngMaster {
+            rng: rng,
+            devices: List::new(),
+            inflight: OptionalCell::empty(),
+        }
+    }
+
+    // TODO: return type is a hacky way of surfacing the return code from rng.get()
+    fn do_next_op(&self) -> ReturnCode {
+        if self.inflight.is_none() {
+            let mnode = self
+                .devices
+                .iter()
+                .find(|node| node.operation.get() != Op::Idle);
+            mnode.map(|node| {
+                let op = node.operation.get();
+                // Need to set idle here in case callback changes state
+                node.operation.set(Op::Idle);
+                match op {
+                    Op::SetClient => {
+                        self.rng.set_client(node);
+                        ReturnCode::SUCCESS
+                    }
```

Comment:
What ensures that the correct client is set when the `Op::Get` executes?

Reply:
Just modified the implementation for `Op::Get` to set the correct client prior to every call